### PR TITLE
[feature-wip](nereids) Implement using join

### PR DESCRIPTION
**Pull request by Kikyou1997**
# Proposed changes

Issue Number: close #xxx

## Problem summary

Describe your changes.

## Checklist(Required)

1. Does it affect the original behavior: 
    - [ ] Yes
    - [ ] No
    - [ ] I don't know
2. Has unit tests been added:
    - [ ] Yes
    - [ ] No
    - [ ] No Need
3. Has document been added or modified:
    - [ ] Yes
    - [ ] No
    - [ ] No Need
4. Does it need to update dependencies:
    - [ ] Yes
    - [ ] No
5. Are there any changes that cannot be rolled back:
    - [ ] Yes (If Yes, please explain WHY)
    - [ ] No

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 35.08 seconds
 load time: 683 seconds
 storage size: 17123273149 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20221223083647_clickbench_pr_67764.html

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/BindSlotReference.java [hunk 1]
```diff
@@ -122,41 +124,60 @@ public List<Rule> buildRules() {
...
+                        UsingJoin using = ctx.root;
+                        LogicalJoin lj = new LogicalJoin(using.getJoinType() == JoinType.CROSS_JOIN
+                                ? JoinType.INNER_JOIN : using.getJoinType(),
+                                using.getHashJoinConjuncts(),
+                                using.getOtherJoinConjuncts(), using.getHint(), (Plan) using.left(),
+                                (Plan) using.right());
+                        List<Expression> unboundSlots = lj.getHashJoinConjuncts();
+                        Set<String> slotNames = new HashSet<>();
+                        List<Slot> leftOutput = new ArrayList<>(lj.left().getOutput());
+                        Collections.reverse(leftOutput);
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/BindSlotReference.java:139]: why need reverse?

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/BindSlotReference.java [hunk 2]
```diff
@@ -122,41 +124,60 @@ public List<Rule> buildRules() {
...
+                        List<Expression> leftSlots = new ArrayList<>();
+                        Scope scope = toScope(leftOutput.stream()
+                                .filter(s -> !slotNames.contains(s.getName()))
+                                .peek(s -> slotNames.add(s.getName())).collect(
+                                        Collectors.toList()));
+                        for (Expression unboundSlot : unboundSlots) {
+                            Expression expression = new SlotBinder(scope, lj, ctx.cascadesContext).bind(unboundSlot);
+                            leftSlots.add(expression);
+                        }
+                        slotNames.clear();
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/BindSlotReference.java:149]: What is the purpose of slotNames?

File: fe/fe-core/src/main/java/org/apache/doris/nereids/stats/JoinEstimation.java [hunk 3]
```diff
@@ -127,7 +127,6 @@ public static StatsDeriveResult estimate(StatsDeriveResult leftStats, StatsDeriv
             rowCount = estimateFullOuterJoin(leftStats, rightStats, join);
         } else {
             LOG.warn("join type is not supported: " + joinType);
-            throw new RuntimeException("joinType is not supported");
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/stats/JoinEstimation.java:130]: why remove this?

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/logical/UsingJoin.java [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+/**
+ * select col1 from t1 join t2 using(col1);
+ */
+public class UsingJoin<LEFT_CHILD_TYPE extends Plan, RIGHT_CHILD_TYPE extends Plan>
+        extends LogicalBinary<LEFT_CHILD_TYPE, RIGHT_CHILD_TYPE> {
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/logical/UsingJoin.java:41]: should implement `org.apache.doris.nereids.trees.plans.algebra.Join`

Comment (Kikyou1997) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/logical/UsingJoin.java:41]: Using Join is actually not any type of join.

Comment (morrySnow) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/logical/UsingJoin.java:41]: in the current implement, it is indeed a Join algebra expression as LogicalJoin

**Review by morrySnow — APPROVED** [review]

**Pull request merged**